### PR TITLE
feat(secrets): track complete file deletion and renaming

### PR DESCRIPTION
**Pull request by lirshindalman**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

add Track complete file deletion and renaming
and add the data in the detectors about is_added and is_removed

result of this PR:
if we choose to scan the git history we will get a secret with added_commit_hash and removed_commit_hash. 
the scan will be a normal scan or a git history scan.
## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/secrets/plugins/custom_regex_detector.py [hunk 1]
```diff
@@ -70,6 +70,8 @@ def analyze_line(
             line_number=line_number,
             context=raw_context,
             is_multiline=False,
+            is_added=is_added,
+            is_removed=is_removed,
```
Comment (nimrodkor) [inline checkov/secrets/plugins/custom_regex_detector.py:74]: @lirshindalman Why does our custom detector need to know this? Can't it be abstracted away in the runner only?

**Review by nimrodkor — COMMENTED** [review]
@lirshindalman What is the result of this PR? Do we have more fields in the report for secrets or something more?

File: checkov/secrets/scan_git_history.py [hunk 2]
```diff
@@ -28,134 +30,149 @@ class EnrichedPotentialSecret(TypedDict):
...
                     break
-    except KeyError:
-        logging.error(f"No added secret commit found for secret in file {file_name}.")
+        except KeyError:
+            logging.error(f"No added secret commit found for secret in file {file_name}.")
+
+    def handle_renamed_file(self, rename_from: str,
+                            rename_to: str,
+                            commit_hash: str) -> None:
+        temp_map: Dict[str, List[EnrichedPotentialSecret]] = {}
```
Comment (maxamel) [inline checkov/secrets/scan_git_history.py:164]: I'd change this name to something reflecting that we accumulate new secrets here

File: checkov/secrets/scan_git_history.py [hunk 3]
```diff
@@ -28,134 +30,149 @@ class EnrichedPotentialSecret(TypedDict):
...
+class GitHistorySecretStore:
+    def __init__(self) -> None:
+        self.secrets_by_file_value_type: Dict[str, List[EnrichedPotentialSecret]] = {}
+
+    def set_secret_map(self, file_results: List[PotentialSecret], file_name: str, commit_hash: str) -> None:
+        # First find if secret was moved in the file
+        equal_secret_in_commit: Dict[str, List[str]] = defaultdict(list)
+        for secret in file_results:
+            secret_key = f'{file_name}_{secret.secret_hash}_{secret.type.replace(" ", "-")}'
+            equal_secret_in_commit[secret_key].append('added' if secret.is_added else 'removed')
```
Comment (maxamel) [inline checkov/secrets/scan_git_history.py:121]: We are not checking is_removed here before setting to removed?

Comment (lirshindalman) [reply checkov/secrets/scan_git_history.py:121]: yes because every line in the diff is added or removed so 'not added' is 'removed'

**Review by maxamel — APPROVED** [review]

File: checkov/secrets/runner.py [hunk 4]
```diff
@@ -458,3 +458,20 @@ def _modify_invalid_secrets_check_result_to_skipped(report: Report) -> None:
...
+        in the next step we will add the connection and the missing data
+        The key is built like this:
+        '{added_commit_hash}_{removed_commit_hash or the string SECRET_NOT_BEEN_REMOVED
+        if the secret not been removed}_{file_name}'
+        """
+        if not enable_git_history_secret_scan:
+            return None, None
+        split_key = key.split('_')
+        added_commit_hash = split_key[0]
+        removed_commit_hash = split_key[1] if split_key[1] != GIT_HISTORY_NOT_BEEN_REMOVED else None
```
Comment (pazbechor) [inline checkov/secrets/runner.py:476]: split_key WILL ALWAYS have at least 2 entries?

File: checkov/secrets/scan_git_history.py [hunk 5]
```diff
@@ -28,134 +30,151 @@ class EnrichedPotentialSecret(TypedDict):
...
+    def __init__(self, root_folder: str, secrets: SecretsCollection, timeout: int = 43200):
+        self.root_folder = root_folder
+        self.secrets = secrets
+        self.timeout = timeout
+        self.secret_store = GitHistorySecretStore()
+
+    def scan_history(self) -> bool:
+        """return true if the scan finished without timeout"""
+        # mark the scan to finish within the timeout
+        with stopit.ThreadingTimeout(self.timeout) as to_ctx_mgr:
```
Comment (pazbechor) [inline checkov/secrets/scan_git_history.py:43]: this is your change? not omri's?

**Review by pazbechor — APPROVED** [review]
see comments, approve for not blocking

File: checkov/secrets/scan_git_history.py (hunk 5 above)
Comment (lirshindalman) [reply checkov/secrets/scan_git_history.py:43]: I modify all file so it's marked as new

File: checkov/secrets/runner.py (hunk 4 above)
Comment (lirshindalman) [reply checkov/secrets/runner.py:476]: if ita git history and we get to this function so yes. 
i set the key in the scan_git_history.py line 68 'key = f'{secret_data["added_commit_hash"]}_{removed}_{secret_data["potential_secret"].filename}''

**Pull request merged**